src/extra/components/adq.py

```diff
@@ -0,0 +1,1302 @@
+        it from every Nth pixel across the entire trace. This will change
+        the native data type to float32 if none is specified explicitly
+        or an output array passed. Note that this type of common mode
+        correction will often not yield useful results for integer types.
+
+        Args:
+            data (ArrayLike): Input data to preprocess, will be
+                converted to np.ndarray currently.
+            out (ArrayLike, optional): Result array, a new is allocated
+                if omitted.
+            cm_period (int, optional): Apply common mode correction with
+                specified sample periodicity if greater than 0, by
+                default 8 or 16 when interleaved.
+            baseline (slice or ArrayLike, optional): Contiguous 1D slice
+                of the trace of each train to determine common mode or
+                baseline or direct baseline data to use, :1000 by default.
+            baselevel (number, optional): ADU value to pull the baseline
+                to, None by default. Note that common mode corrections
+                if enabled always pull the baselevel to zero unless
+                specified otherwise here.
+
+        Returns:
+            out (ndarray): Preprocessed input data, converted to ndarray
+                if a different array was passed.
+        """
+
+        if cm_period < 1:
+            raise ValueError('Common mode must be positive number')
+
+        out = np.zeros_like(data, dtype=np.float32)
+
+        self._correct_cm_by_train(data, out, cm_period, baseline, baselevel)
+
+        return out
+
+    def pull_to_baselevel(self, data, baseline, baselevel):
+        out = np.zeros_like(data, dtype=np.float32)
+
+        self._pull_to_baselevel(data, out, baseline, baselevel)
+
+        return out
+
+    def reshape_to_pulses(self, data, pulse_ids, first_pulse_offset=None):
+        """Reshape train data to pulse data.
+
+        This method performs pulse separation by splitting the trace
+        acquired by train into individual traces by pulse based on the
+        pulse information the component is initialized with.
+
+        Args:
+            data (ArrayLike): Digitizer trace(s) for one or more trains,
+                last axis is assumed to be sample axis and on a train
+                boundary.
+            pulse_ids (ArrayLike, optional): Pulse IDs of a single train.
+            first_pulse_offset (int, optional): Sample where the first
+                pulse begins, by default the value the component was
+                initialized with.
+
+        Returns:
+            out (ArrayLike): Reshaped pulse traces.
+        """
+
+        # TODO: Support data.ndim > 2
+
+        if not isinstance(data, np.ndarray):
+            data = np.asarray(data)
+
+        if first_pulse_offset is None:
+            first_pulse_offset = self._first_pulse_offset
+
+        pulse_shape = data.shape[:-2] + (
```

Comment:
Should this be 

```suggestion
        pulse_shape = data.shape[:-1] + (
```

? If you're starting with a 2D array, `[:-2]` gets you an empty tuple.

Reply:
That is intentional. The method is supposed to work for `data.ndim > 2` at some point, and this achieves that already. But I didn't test it yet, hence the `# TODO`. Frankly I do not expect someone trying often, the idea is to support arbitrary sorting on the user side with the last axes being `(..., num_trains, samples)`.

Reply:
I see, but in that case, I think there's something else wrong. `pulse_ids` is documented as 'Pulse IDs of a single train', but `len(pulse_ids)` is used for the output shape. Shouldn't it be `len(pulse_ids) * n_trains` ?

Reply:
Yes, the docstring is wrong due to copypasta. Well spotted. It must be all pulse IDs, as it explicitly supported different number of pulses.

Reply:
But actually, I intended to use the internal pulse information, so there's also a mistake in the implementation. This can be relaxed later, but for now the methods only support the data the component also returns.

Reply:
Thanks, that makes sense now.